frontend/app/controllers/Promotions.scala

```diff
@@ -65,13 +69,18 @@ object Promotions extends Controller {
       }
 
     val promoCode = PromoCode(promoCodeStr)
+    val notFound = NotFound(views.html.error404())
 
-    (for {
-      promotion <- TouchpointBackend.Normal.promoService.findPromotion(promoCode)
-      html <- if (promotion.expires.isBeforeNow) None else findTemplateForPromotion(promoCode, promotion, request.path)
-    } yield Ok(html).withCookies(sessionCookieFromCode(promoCode)))
-      .getOrElse(NotFound(views.html.error404()))
-
+    //Use a \/ disjunction in this for comprehension, so we don't just return a None
+    (
+      for {
+        promotion <- TouchpointBackend.Normal.promoService.findPromotion(promoCode) \/> notFound
+        _ <- if (promoCodeStr.toUpperCase != promoCodeStr) \/.left(Redirect("/p/" + promoCodeStr.toUpperCase)) else \/.right(Unit)
```

Comment:
Some of these lines are very long and the code is hard to read. Is there any way it can be simplified? any more intermediate variables that can be used?